drivers/sensor/aht20/aht20.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2022 Rafael Lee <rafaellee.img@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT asair_aht20
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/crc.h>
+
+#define AHT20_STATUS_LENGTH	1
+#define AHT20_TOTAL_READ_LENGTH 7
+#define AHT20_CRC_POLY		0x31 /* polynomial 1 + x^4 + x^5 + x^8 */
+#define AHT20_CRC_INIT		0xff
+
+#define AHT20_CMD_RESET		     0xBA
+#define AHT20_CMD_TRIGGER_MEASURE    0xAC
+#define AHT20_TRIGGER_MEASURE_BYTE_0 0x33
+#define AHT20_TRIGGER_MEASURE_BYTE_1 0x00
+#define AHT20_CMD_GET_STATUS	     0x71
+#define AHT20_CMD_INITIALIZE	     0xBE
+
+#define AHT20_FULL_RANGE_BITS	 20
+#define AHT20_TEMPERATURE_RANGE	 200
+#define AHT20_TEMPERATURE_OFFSET 50.0
+
+typedef union {
+	struct {
+		uint8_t: 3;		/* bit [0:2] reserved */
+		uint8_t calibrated: 1;	/* bit [3] calibrated */
+		uint8_t: 1;		/* bit [4] reserved */
+		uint8_t: 2;		/* bit [5:6], AHT20 datasheet v1.1 removed 2 mode bits */
+		uint8_t busy: 1;	/* bit [7] reserved */
+	};
+	uint8_t all;
+} __attribute__((__packed__)) aht20_status;
+
+struct aht20_data {
+	struct i2c_dt_spec bus;
+	uint32_t humidity;
+	uint32_t temperature;
+};
+
+LOG_MODULE_REGISTER(AHT20, CONFIG_SENSOR_LOG_LEVEL);
+
+static int aht20_channel_get(const struct device *dev, enum sensor_channel chan,
+			     struct sensor_value *val)
+{
+	struct aht20_data *drv_data = dev->data;
+	float temperature;
+
+	switch (chan) {
+	case SENSOR_CHAN_HUMIDITY:
+		val->val1 = 0;
+		val->val2 = (drv_data->humidity) * 1000000LL / (1 << AHT20_FULL_RANGE_BITS);
+		break;
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		temperature = (float)drv_data->temperature / (1 << AHT20_FULL_RANGE_BITS) *
+				      AHT20_TEMPERATURE_RANGE -
+			      AHT20_TEMPERATURE_OFFSET;
+		val->val1 = (int32_t)temperature;
+		val->val2 = (temperature - val->val1) * 1000000LL;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+	return 0;
+}
+
+static int aht20_init(const struct device *dev)
+{
+	struct aht20_data *drv_data = dev->data;
+	aht20_status sensor_status = {0};
+	uint8_t const inquire_status_seq[] = {AHT20_CMD_GET_STATUS};
+	uint8_t const initialize_seq[] = {AHT20_CMD_INITIALIZE};
+	int ret = 0;
+
+	k_sleep(K_MSEC(40)); /* wait for 40ms */
+	ret = i2c_write_read_dt(&drv_data->bus, inquire_status_seq, sizeof(inquire_status_seq),
+				&sensor_status.all, AHT20_STATUS_LENGTH);
+	if (ret) {
+		LOG_ERR("Inquire status failed");
+	} else {
+		LOG_INF("Inquire status successfully");
+	}
+	if (!sensor_status.calibrated) {
+		LOG_INF("Calibrated bit is 0, need initialization");
+		ret = i2c_write_dt(&drv_data->bus, initialize_seq, sizeof(initialize_seq));
+		if (ret) {
+			LOG_ERR("Initialization filaed");
+		} else {
+			LOG_INF("Initialization successful");
+		}
+		k_sleep(K_MSEC(10));
```

Comment:
How was the sleep duration chosen?

Reply:
Described in AHT20 datasheet v1.1

https://files.seeedstudio.com/wiki/Grove-AHT20_I2C_Industrial_Grade_Temperature_and_Humidity_Sensor/AHT20-datasheet-2020-4-16.pdf

5.4 Sensor reading process

1. Wait **40ms** after power-on. Before reading the temperature and humidity values, ﬁrst check whether the calibration enable bit Bit [3] of the status word is 1 (you can get a byte of status word by sending 0x71). If not 1, need to send 0xbe command (for initialization), this command parameter has two bytes, the ﬁrst byte is 0x08, the second byte is 0x00, and then wait for **10ms**.